commands/provider/marketplace_provider.go

```diff
@@ -0,0 +1,428 @@
+	}
+	var outputData SendSignedTransactionTaskSuccessOutput
+	return &outputData, decodeOutput(outputKey, outputDataS, &outputData)
+}
+
+// IsAuthorized executes the task send signed transaction.
+func (p *MarketplaceProvider) IsAuthorized(sidHash string) (bool, error) {
+	outputKey, outputDataS, err := p.executeTaskAndWaitResult(MarketplaceServiceID, "isAuthorized", &IsAuthorizedTaskInputs{
+		SidHash: sidHash,
+	})
+	if err != nil {
+		return false, err
+	}
+	var outputData IsAuthorizedTaskSuccessOutput
+	if err = decodeOutput(outputKey, outputDataS, &outputData); err != nil {
+		return false, err
+	}
+	return outputData.Authorized, nil
+}
+
+// PublishDefinitionFile upload and publish the tarball and definition file and returns the address of the definition file
+func (p *MarketplaceProvider) PublishDefinitionFile(path string) (string, error) {
+	ipfs := ipfs.New()
+	tar, err := archive.TarWithOptions(path, &archive.TarOptions{
+		Compression: archive.Gzip,
+	})
+	if err != nil {
+		return "", err
+	}
+	tarballResponse, err := ipfs.Add("tarball", tar)
+	if err != nil {
+		return "", err
+	}
+
+	definitionFile, err := p.createDefinitionFile(path, tarballResponse.Hash)
+	if err != nil {
+		return "", err
+	}
+
+	definitionResponse, err := ipfs.Add("definition", bytes.NewReader(definitionFile))
+	if err != nil {
+		return "", err
+	}
+	return definitionResponse.Hash, nil
+}
+
+func (p *MarketplaceProvider) createDefinitionFile(path string, tarballHash string) ([]byte, error) {
+	definition, err := importer.From(path)
+	if err != nil {
+		return nil, err
+	}
+	var data ManifestData
+	data.Version = PublishVersion
+	data.Service.Deployment.Type = DeploymentType
+	data.Service.Deployment.Source = tarballHash
+	data.Readme, err = readme.Lookup(path)
+	if err != nil {
+		return nil, err
+	}
+	data.Definition = definition
+	return json.Marshal(data)
+}
+
+// TODO: to refactor with the same function in service_provider
+func (p *MarketplaceProvider) executeTaskAndWaitResult(sid, taskKey string, inputData interface{}) (string, string, error) {
+	var (
+		result *coreapi.ResultData
+		err    error
+		data   []byte
+		tags   = []string{uuid.NewV4().String()}
+	)
+	data, err = json.Marshal(inputData)
+	if err != nil {
+		return "", "", err
+	}
+	// TODO: the stream should be terminated when one result received
+	listenResultsC, resultsErrC, err := p.listenResults(sid, taskKey, "", tags)
+	if err != nil {
+		return "", "", err
+	}
+	err = p.executeTask(sid, taskKey, string(data), tags)
+	if err != nil {
+		return "", "", err
+	}
+	select {
+	case result = <-listenResultsC:
+	case err = <-resultsErrC:
+	}
+	if err != nil {
+		return "", "", err
+	}
+	if result.Error != "" {
+		return "", "", errors.New(result.Error)
+	}
+	return result.OutputKey, result.OutputData, nil
+}
+
+// listenEvents returns a channel with event data streaming.
+func (p *MarketplaceProvider) listenEvents(id, eventFilter string) (chan *coreapi.EventData, chan error, error) {
```

Comment:
All these functions should be extracted later in a cli helper package for example